runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/DfsArtifactRetrievalService.java

```diff
@@ -0,0 +1,135 @@
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi;
+import org.apache.beam.model.jobmanagement.v1.ArtifactRetrievalServiceGrpc;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.ResourceId;
+
+/**
+ * An {@link ArtifactRetrievalService} that uses distributed file systems as its backing storage.
+ */
+public class DfsArtifactRetrievalService
+    extends ArtifactRetrievalServiceGrpc.ArtifactRetrievalServiceImplBase
+    implements ArtifactRetrievalService {
+
+  private static int artifactChunkSizeBytes = 2000000; // 2MB
+
+  public static DfsArtifactRetrievalService create() {
+    return new DfsArtifactRetrievalService();
+  }
+
+  @Override
+  public void getManifest(
+      ArtifactApi.GetManifestRequest request,
+      StreamObserver<ArtifactApi.GetManifestResponse> responseObserver) {
+    try {
+      ArtifactApi.ProxyManifest proxyManifest = loadManifestFrom(request.getRetrievalToken());
+      ArtifactApi.GetManifestResponse response =
+          ArtifactApi.GetManifestResponse
+              .newBuilder()
+              .setManifest(proxyManifest.getManifest())
+              .build();
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void getArtifact(
+      ArtifactApi.GetArtifactRequest request,
+      StreamObserver<ArtifactApi.ArtifactChunk> responseObserver) {
+    try {
+      ArtifactApi.ProxyManifest proxyManifest = loadManifestFrom(request.getRetrievalToken());
+      // validate that name is contained in manifest and location list
+      boolean containsArtifactName =
+          ImmutableList.copyOf(proxyManifest.getManifest().getArtifactList())
+              .stream()
+              .anyMatch(metadata -> metadata.getName().equals(request.getName()));
+      if (!containsArtifactName)  {
+        throw new ArtifactNotFoundException(request.getName());
+      }
+      // look for file at URI specified by proxy manifest location
+      ImmutableList<ArtifactApi.ProxyManifest.Location> locationList =
+          ImmutableList.copyOf(proxyManifest.getLocationList());
+      ArtifactApi.ProxyManifest.Location location =
+          locationList
+              .stream()
+              .filter(loc -> loc.getName().equals(request.getName()))
+              .findFirst()
+              .orElseThrow(() -> new ArtifactNotFoundException(request.getName()));
+      ResourceId artifactResourceId =
+          FileSystems.matchNewResource(location.getUri(), false /* is directory */);
+      ReadableByteChannel artifactByteChannel = FileSystems.open(artifactResourceId);
+      ByteBuffer byteBuffer = ByteBuffer.allocate(artifactChunkSizeBytes);
+      while (artifactByteChannel.read(byteBuffer) > -1) {
+        ByteString data = ByteString.copyFrom(byteBuffer);
```

Comment:
Don't you need to flip the buffer before reading it?

Reply:
Huh, yes I do.  Thanks.